Pw2wann workchain

DanielMarchand opened the pull request:
Add a SplitPw2wann WorkChain to help parallelize that step

Commit:
added pw2wann workchain

Commit:
working version of SplitPw2Wann

Commit:
minor revisions

Commit:
updated tb-opt workchain to use SplitPw2wann workchain

greschd commented on aiida_tbextraction/fp_run/wannier_input/pw2wann.py:
```diff
@@ -0,0 +1,409 @@
+        # Writing mmn file, unfortunately this is a bit messy because
+        # write_mmn dumps to a file which must be read again later
+        mmn_filename = "aiida.mmn"
+        with tempfile.TemporaryDirectory() as tmp_dir:
```
AFAICT we only use a single file from the `TemporaryDirectory` here - meaning we could also use `NamedTemporaryFile` instead.

DanielMarchand replied:
added in latest commit

greschd commented on aiida_tbextraction/fp_run/wannier_input/pw2wann.py:
```diff
@@ -0,0 +1,409 @@
+            del mmn_collected
+            gc.collect()  # for paranoia's sake run garbage collection
+
+            with open(output_file, mode='r') as f_in:
```
Here we could use `FolderData.put_object_from_file` to copy the file (probably, that'll be more efficient).

DanielMarchand replied:
added in latest commit

greschd commented on aiida_tbextraction/fp_run/wannier_input/pw2wann.py:
```diff
@@ -0,0 +1,409 @@
+                with folder.open(mmn_filename, mode='w') as f_out:
+                    f_out.write(f_in.read())
+
+        # writing amn, eig (and any other) outputs
```
I think the `(and any other)` part of this comment should be a TODO: Currently, other files are _not_ copied. 

We'll probably also need some logic to determine _which_ files should be retrieved.

DanielMarchand replied:
Correct, let's discuss

greschd replied:
I don't think this needs doing right now, just adding a TODO comment stating what is currently retrieved and that other output files aren't is enough.

greschd commented on aiida_tbextraction/fp_run/wannier_input/pw2wann.py:
```diff
@@ -0,0 +1,409 @@
+        no_mmn_calc = self.ctx['pw2wann_no_mmn']
+        r_folder = no_mmn_calc.outputs.remote_folder
+        for filename in filenames:
+            with tempfile.TemporaryDirectory() as tmp_dir:
```
Here, I'm not sure if we should put all the files into the same `TemporaryDirectory` (switching the loop order), and then use `FolderData.put_object_from_tree` to copy the whole directory to the repository at once.

That'd use more disk space (temporarily), but IMO the code would be a bit simpler.

greschd replied:
@DanielMarchand opinions on this? We can also leave it, up to you.

greschd reviewed (commented):
Thanks @DanielMarchand, looks good overall - I added some minor comments / suggestions.

I haven't yet ran the tests locally, but will do so to check if there are any changes needed there (on the QE part that doesn't run in CI). On a related note, I'm not sure why CI didn't run at all here...

greschd commented:
Regarding the merge conflicts: I'm fairly certain that `exclude=['pw.structure', 'kpoints']` (the version in `develop`) is the correct one.

For the second merge conflict (from the `UNK*` PR https://github.com/greschd/aiida-tbextraction/pull/19) you can just add that to the TODO comment for retrieving additional files from the "no MMN" calculation.

Commit:
Add wannier90-utils and more_itertools to dependencies.

greschd commented:
> On a related note, I'm not sure why CI didn't run at all here...

I took the liberty to rebase the commits, to make the CI run. The problem was that this branch didn't contain the changes to move from Travis to Github Actions.

greschd commented on aiida_tbextraction/fp_run/wannier_input/pw2wann.py:
```diff
@@ -0,0 +1,409 @@
+import copy
+import gc
+import io
+import more_itertools
```
`more_itertools` is not in the standard library, this belongs in the second block of imports.

DanielMarchand replied:
corrected in latest commit

greschd commented on aiida_tbextraction/fp_run/wannier_input/pw2wann.py:
```diff
@@ -0,0 +1,409 @@
+import numpy as np
+from w90utils.io import write_mmn
+
+from collections import ChainMap
```
`collections` _is_ in the standard library, this belongs in the first block of imports.

DanielMarchand replied:
corrected in latest commit

Commit:
Fix exclude from NSCF PwBaseWorkChain

greschd commented:
I think the tests also need some changes, for me `tests/fp_run/test_qe_fp_run.py` fails.

Commit:
Made corrections to pw2wann part of tb-opt tests

greschd commented on tests/conftest.py:
```diff
@@ -263,8 +263,12 @@ def inner():
                 'pw2wannier': {
-                    'code': code_pw2wannier90,
-                    'metadata': get_metadata_singlecore()
+                    # Use a huge batchsize to force it all at once
```
Shouldn't there be at least one test where we actually use more than one batch?

DanielMarchand replied:
Yeah that would make sense

greschd replied:
Fixed in 598e5d1

Commit:
addressed most PR comments

DanielMarchand commented:
> I think the tests also need some changes, for me `tests/fp_run/test_qe_fp_run.py` fails.
Working again

greschd commented on aiida_tbextraction/fp_run/wannier_input/_qe.py:
```diff
@@ -13,7 +13,7 @@
 from aiida_wannier90.calculations import Wannier90Calculation
 from aiida_quantumespresso.workflows.pw.base import PwBaseWorkChain
-from aiida_quantumespresso.calculations.pw2wannier90 import Pw2wannier90Calculation
+from aiida_tbextraction.fp_run.wannier_input.pw2wann import SplitPw2wannier90
```
For consistency, we should name the file `pw2wannier` (sorry, didn't spot that one last time around).

Commit:
Minor changes to pw2wannier workchain.

- Rename file to 'pw2wannier90'
- Add TODO for 'full' pw2wannier90.x input, retrieving all files
- Use a single `TemporaryDirectory` for retrieving no-MMN output

Commit:
Add test for 'batched' mode in split pw2wannier, fix 'unbatched'.

Add a test for 'batch size < number of target bands' in the
split pw2wannier90 workflow. This uncovered an error in the
opposite case (where only a single MMN calculation is run), by
comparing the resulting MMN matrices, which is also fixed.

Commit:
Comment out VASP - related code.

There is a dependency conflict between pytest>=6.0 and aiida-vasp.
Because we don't actually have a working VASP implementation at
the moment, we can just comment out all related code and remove
the aiida-vasp dependency to temporarily resolve this.
